### ISSUE TITLE
Simple, minimal server-side support

### ISSUE THREAD

**Issue by Coises** [keeweb/keeweb#1737]
KeeWeb can be used with a few different proprietary remote storage schemes, but the only non-proprietary standard is WebDAV.  I think the reason is that WebDAV appears to be the only non-proprietary standard that allows client-side ECMAScript in a browser to update a file on a server without application-specific server-side support.

Alas, support for the WebDAV standard is not always so standard.  Support on shared hosts is spotty, and if there is a simple way to add it, I couldn't find it.  Even where it is supported, other confusions (CORS seems to be a common one) cause failures.  I've gotten it to work by installing NextCloud, which feels like swatting a fly with a sledgehammer.  (At last check, #492 still applies.  This is probably a NextCloud absurdity, not a KeeWeb problem... but it demonstrates how completely flaky support for the WebDAV "standard" really is.)

Would it be possible to create a _simple_ server-side service specifically to support KeeWeb that could be installed easily on any server or shared hosting account that supports PHP?

I'm thinking that some clever use of challenge-response should make it possible to use the same passphrase that encrypts the KeePass file to protect it from access/vandalism as well, so one wouldn't have to remember separate login credentials to access the file.  For example, when a file is to be saved, in addition to the file, KeeWeb would send (over https) a randomly-generated string and the result of encrypting that same string with the passphrase.  The server-side service would store both of these in a file of the same name with a different extension.  When the KeePass file is next requested, the server would send the encrypted string and the client would respond with the unencrypted string, proving that the user supplied the correct passphrase.  Creating a new KeePass file (other than the first one) would require having already successfully opened one file.  No user name would be necessary; the file name itself would serve that role.

I don't know how broad the use for this would be, or how difficult it would be to build such a thing.  It's a shame SFTP isn't viable, but as I understand it, that won't work from a browser.

**Comment by antelle** [comment]
Hi!

I'm not sure if there's demand for a simple web server, but I see why it may make sense. I'd like to get some feedback on what is desired and who would prefer setting it up this way. It should probably support more languages rather than PHP, for example, it can be also in node.js, which is quite popular now.

One way it can be implemented without a lot of effort, is using WebDAV protocol, but pre-configured, so that we have CORS headers and so on. But maybe not, depending on how people would like to authenticate there and what features we want to support, simple REST API may be more appropriate. It's better to authenticate via OAuth rather than how it's done in most of WebDAV setups.

It would be also great if this server exposes other API, for example, it may act as an FTP or S3 server for KeePass KeePassXC. It makes sense to explore which sync options they provide, so that people are not locked into KeeWeb setup and can use other clients as well.

Another interesting feature that the webserver can support is hosting KeeWeb right there, if the user prefers. So that you installed it you can access your file via browser.

Potentially this can also be extended for adding team features in future, but now it's not clear how.

**Comment by Coises** [comment]
> One way it can be implemented without a lot of effort, is using WebDAV protocol, but pre-configured, so that we have CORS headers and so on.

I'm thinking of cases where WebDAV isn't already available. For example, Pair Networks, who provide my web space, don't enable Apache WebDAV on shared hosting accounts. I can run PHP scripts, and cron jobs at a maximum of once per hour. I don't have root access. I could have all that with a virtual private server, but at three times the cost and a lot more administrative responsibility.

But if you mean an easy way to add WebDAV to a typical shared hosting account from a provider that doesn't support it directly, that would be a big step forward, and potentially useful for other things besides KeeWeb. I'm just starting to wonder how much anyone should trust WebDAV at all... it seems "flaky" in practice, like too many twists have been grafted onto an old protocol from a simpler time. If a modern, non-proprietary replacement exists that can be used from client-side ECMAScript in a browser, that would be wonderful.

> But maybe not, depending on how people would like to authenticate there and what features we want to support, simple REST API may be more appropriate. It's better to authenticate via OAuth rather than how it's done in most of WebDAV setups.

We will always need a really strong yet memorable passphrase for the KeeWeb database. If we have to authenticate independently to get to the database, that's more credentials we have to remember — we obviously couldn't retrieve them from the database when we needed to access the database. Knowing the URL of our service, selecting the database file from a list (or providing a user name, if you want to cover use cases where one host is supporting many users) and knowing the database passphrase ought to be enough. That's why I suggested some kind of challenge-response in which the client proves that it knows the passphrase before being granted access to the database. (It wouldn't _have_ to be impossible to retrieve the encrypted database without knowing the passphrase, since the would-be hacker still couldn't read it... but then they could employ offline cracking techniques, so it's better if they can't even download it. Rate limiting for challenge-response attempts would help thwart brute force online cracking attempts. In any case, to prevent vandalism, it can't be possible to update the database without either proving knowledge of the passphrase or authenticating in some other way.)

> It would be also great if this server exposes other APIs, for example, it may act as an FTP or S3 server for KeePass or KeePassXC.

I think almost any hosting service that allows the owner run a custom script (PHP or anything else) is going to have FTP and probably SFTP access — how else would the user upload the script? If KeeWeb itself could use SFTP, I'd be happy... but my understanding is that you can't use SFTP, or even FTP, from client ECMAScript in a web browser.

> Another interesting feature that the webserver can support is hosting KeeWeb right there, if the user prefers. So that you installed it you can access your file via browser.

Yes, I think that's an important use case. I do that now — the config.json file points directly to the NextCloud WebDAV URLs of my databases, so if I'm stuck away from home, I only have to remember the URL of my KeeWeb page, my database passphrases and (unfortunately) my NextCloud user name and passphrase. It would just be nice to have something dead simple, reliable and to the point. NextCloud is a big, complex thing with too many potential points of failure when all I need is to make two small files available on demand to KeeWeb.

**Comment by antelle** [comment]
> But if you mean an easy way to add WebDAV to a typical shared hosting account from a provider that doesn't support it directly, that would be a big step forward, and potentially useful for other things besides KeeWeb

Yes, I mean that, because webdav is implemented in many libraries and it should be simple to add it

> I'm just starting to wonder how much anyone should trust WebDAV at all...

Well, KeeWeb is not using a lot of WebDAV features, what usually goes wrong in KeeWeb, is CORS configuration, most of servers don't support it for obvious reasons and it's hard to understand and configure in general especially for unprepared people. That's why I was thinking, maybe if this server provides WebDAV API with all necessary configuration, it will work.

> non-proprietary replacement exists that can be used from client-side ECMAScript in a browser, that would be wonderful

There's just REST API, we can use it instead. Maybe it will be easier this way. However it needs CORS as well, if KeeWeb is hosted on another domain. That's actually another question, where would people like to see the script being described here and KeeWeb app, on the same host or different?

**Comment by Coises** [comment]
My thought is that the more common use case would be all on the same host; if you have the rights to upload a script, you certainly have the rights to upload the KeeWeb web page. Having the whole thing as a single, ready-to-go package would be wonderful. I'm sure there would be cases, though, where someone would want to separate the storage from the web-facing interface.

It seems like all CORS checking could just be turned off (if that's even possible these days); the whole point of KeePass and KeeWeb is that even if someone gets the database file they can't read it without the passphrase, and no matter where you are, you can read it if you have the passphrase and access to a web browser. CORS only protects browser access, anyway; if a desktop client can access the server, anyone could hack it to continue to access the server with whatever modifications they like. The server has to protect itself with no regard to the origin of requests. (But that's always true. CORS — if I understand this correctly — doesn't protect the server, it protects the user from being tricked. I don't know enough about server-side programming to know if this is possible, but I was thinking of a simple, essentially non-stateful server script. One would never "log in" to it; each request-response would stand on its own. No cookies, no concept of a "session." The client-side would be stateful, but not the server.)

By the way... I wonder what [BitWarden](https://github.com/bitwarden/server) does?  (It's not likely that I have the skill/knowledge to understand it.) I think it has a system where the only passphrase you need is the one for the database itself, but I don't know if it's guaranteed zero-knowledge — in my very personal opinion, KeeWeb should not consider anything that does not keep the server side zero-knowledge.

**Comment by antelle** [comment]
BitWarden has their own format and an active server component. This makes a lot of things easier, they don't have to download a file and be compatible with other forks. This way it can offer 2FA and other features. Also it doesn't work without server, while KeeWeb does.

**Comment by dodedodo** [comment]
It is quite easy to set-up an nginx server/vhost to support webdav with HTTP basic-auth. I've done just that for a couple of users. However, there is no self-service for accounts (password reset, account creation etc). If someone is interested in this set-up let me know, I'll drop my config somehwere.

**Do as I say, not as I do.** I'd recommend using something like nextcloud. There is a lot of development put into security. You got 2 factor authentication, app passwords, rate limiting, account creation & recovery, sharing features and a GUI already baked in. There is also a keeweb plugin so you don't even need to set-up a separate web interface. Why re-invent the wheel? 

Keeweb is a client-side electron app (I assume?). Essentially a desktop app using a browser as development framework. Adding a secure server-side component is no trivial task. Although I do agree it would be nice to have keeweb as a SaaS solution, in my opinion it's not worth the effort.

Regarding protocols; SFTP and FTPS are 2 wildly different ways things. The former is essentially a sub-protocol of ssh (much like scp), while the latter is plain ftp over TLS(SSL). 
WebDav isn't obscure at all. It is HTTP with a few new request methods. I also think CORS (just like SSL) should be implemented in a proxy, not in the application itself.

**Comment by MichaelLeeHobbs** [comment]
@antelle I have paid time to setup a nodejs server api for this. This is the run down of what we are looking at.

Nodejs 14 Containerized in an environment where each container can securely talk to each other ie http between containers with a nginx container handling SSL offloading for client traffic. Auth is handled by another container but I can toss that into the keeweb api container as it will be mostly copy paste. As of right now the follow Auth is available 2FA authenticator, mutual TLS, and local. Passport.js is used so it would be easy to add all sorts of auth methods.

What would you wish list for api be?

get:/api/db/:id
post:/api/db
put:/api/db/:id
del:/api/db/:id

along with some auth paths?

While the target is a container there is no reason it will not run outside a container we do that all the time in dev. Let me know your thoughts this is in the planning/thinking about it column and will be a few weeks before it moves into one of our todo queues.

**Comment by antelle** [comment]
@MichaelLeeHobbs I'm not sure about this. If I understand the original idea correctly, it should be a _simple_ script ready to be deployed by people with low-to-none experience in web-dev, not a full-fledged web-app that also needs a rather complex deployment and setup. Not that it's not possible, but it changes the focus from the proposal made here and it becomes something else.

At this point, it's of course too early to say what API should be exposed.

**Comment by MichaelLeeHobbs** [comment]
@antelle I can setup a single executable for win/linux/mac. Config would be a json file or environment variable if one is even needed. Would use a single json file as storage for any persistent data outside the keepass/keeweb db files which would be stored side by side with the json file. For windows can setup a chocolatey install and linux some wget/bash setup as well as an npm install. Have no access to a mac so best I can do is a executable.


> I also think CORS (just like SSL) should be implemented in a proxy, not in the application itself.
I agree with that statement and can assist with nginx config.

Just think about it. Will be a few weeks before I have work time to work on this. Would love to work with you on this as I'm a big fan of opensource. Otherwise I'll get pushed into some close source solution. My boss is willing to let me do this ie you let me know what the api should look like to plug into the frontend you have built. The only thing extra I'd have to do is build a plugin system to plug our auth in over the default auth which is not hard. Default auth being Passportjs based.

**Comment by antelle** [comment]
An executable is most likely not what is expected. But maybe yes, I don't know to be honest what would be the expectation here. Maybe a linux package would work, or a docker image, one way to figure out can be building an MVP and trying.

What do you think about starting a project with it and publishing it, so that people can install your product if they find it useful. The same way as [nextcloud-keeweb](https://github.com/jhass/nextcloud-keeweb) is done, but a completely self-hosted variant, without nextcloud. I don't think you need any support in KeeWeb for it, but I'm ready to support the storage if KeeWeb is opened in a certain way, I assume this will be the only change needed.

There may be different visions and versions of it, I think it would be good for all of us if there are options. Imagination can take us far:
- with or without additional authentication
- with flexible config support
- with rights per file for different users
- with some extra management UI
- with CORS or flexible domain hosting options
and so on.

**Comment by Coises** [comment]
@Antelle,

You are correct, I was thinking of something quite different than the idea @MichaelLeeHobbs has (to the extent that I understand the idea he has, which admittedly is not that much).

I was thinking of something that:

1. Can be installed on a shared hosting account with minimal privileges. (E.g., can run php scripts, but can't change php configuration; can use .htaccess files but otherwise can't change Apache configuration; can use SFTP to upload files, may or may not be able to run SSH commands or establish a terminal session, can't use sudo; in my case, can establish cron jobs, but they can run at most once per hour.)

2. Can be installed by uploading a file (or a set of files) to the directory that will serve as the root of the installation and then (if this is needed) accessing that file using a web browser to complete the installation.

3. Can host KeePass databases for both the desktop and web versions of KeeWeb.

4. Doesn't conflict with allowing other KeePass-compatible clients to access the same database files by other methods (in particular, SFTP, which is usually going to be possible for any shared hosting environment, and is available in KeePass... but unfortunately can't be used in KeeWeb because it can't be used in a browser).

5. Can (but doesn't have to) host a copy of the KeeWeb web client page.

6. Is a "dumb" server in most senses — it responds to GET and POST requests to serve and update the files but knows nothing about the KeePass database structure itself.

7. Uses a stateless, zero-knowledge protocol — some form of challenge/response — to allow KeeWeb clients to authenticate their access to the database using the same passphrase that encrypts the database, so users don't have to commit an additional strong passphrase to memory. Essentially each restricted transaction (GET or POST, I think) would contain any needed authentication, without the server being expected to maintain "sessions" or "logins" or use any cookies (but the server would keep something like a public key associated with each database, where the public/private key pair is generated by the client based on the passphrase).

Note that 1 and 2 are satisfied by NextCloud... but it's a rather big thing to install for something that seems so simple.

Item 4 makes me think that there might be another way to meet every criteria except 7 — some method of creating a "proxy" that would allow a client-side script, like KeeWeb, to read and write files through SFTP.

**Comment by marclaporte** [comment]
I am interested in adding to Tiki Wiki CMS Groupware: https://doc.tiki.org/File-Gallery

I agree WebDAV support is inconsistent, but there is a popular PHP lib, which many projects use: https://doc.tiki.org/SabreDAV

If some PHP code is written for this, it would be nice if this was available from https://packagist.org/ so many projects could converge on this solution.

Thank you!

**Comment by MichaelLeeHobbs** [comment]
@antelle Let me know thoughts. If your good with this I'll toss this on our internal project board.

MVP
Stack: nodejs/express/express session/passportjs/some kind of local storage ie no database to keep this simple

Will provide and API client for you to hook into the following.
Routes:
post:/api/auth
Auth types: none, api key, local, local with 2FA ie google auth

crud:/api/group
simple group management

crud:/api/account
simple group management

crud:/api/file - using https://tus.io/ for file transfers
simple group management
file can belong to a single account or a group

crud:/api/config
simple config management

Targets
npm - ie npm install global - easy self hosting
Docker - ie docker run - easy self hosting if you use docker already
Executables - Downloadable - easy everything but have to make sure you pick the right download for given enviroment
linux - wget ... | bash run - not sure could be a pain
windows - chocolatey (maybe) - not sure, use it all the time but never made a package for it
mac - Download only cause I know zero about mac =)

Docker Compose
- KeeWeb UI
- KeeWeb Server
- Nginx Server - SSL off load, CORS, etc

I know that seems like a lot but we have 90% of that code written, I'd just need to copy/paste and glue it all together. 

For the MVP phase will just do 
1. git clone
2. yarn install
3. yarn start:dev
4. on first run admin with password will be generated and logged to console

Once both sides are happy I'll write the scripts to generate all the targets. Will need help on mac. I can generate an executable but no way to test.

**Comment by antelle** [comment]
@MichaelLeeHobbs I won't give any opinion about the setup because it's up to you and I'll review the part where you will communicate with KeeWeb to load files.

Authentication will be handled on the protocol level, right? So that if you loaded KeeWeb, you're already authenticated and we don't have to do anything else, KeeWeb just needs to load files from where it's said to.

How are you going to obtain the list of files? There are two ways of doing that:
1. you provide a config to KeeWeb where you specify the list of files available to load; KeeWeb displays all of them in the UI
2. you expose some API to load the list of files and KeeWeb displays the list of files based on it; users can choose files from the list and then these files will be displayed on the main page (how for example Dropbox is implemented)

> using https://tus.io/ for file transfers

Can it be just POST/GET? We don't really need anything else more advanced here, and support for POST/GET is already built into KeeWeb in form of WebDAV. Hard parts in the WebDAV setup are is listing and authentication, file loading / saving works perfectly.

I'd like to say it explicitly to avoid confusion. The project will be maintained by you, it will integrate KeeWeb in some way and provide additional services on top of it, such as file hosting as authentication. What I'll help you with, is implementation of the storage in KeeWeb (if it's required).

**Comment by MichaelLeeHobbs** [comment]
@antelle I created a very basic get/put/post/delete express server for storing files. If I get a chance today I will finish testing and fix any bugs and then post it to github.

routes:
/ - get - static content from src/public return is handled by express.static
/healthcheck - get - returns ok to make things like AWS ELB happy plain text 200
/api/file - get - returns a list of files/id's - returns json 200/500
/api/file/:id - get - returns a file 200/404/500
/api/file - post - store a file - returns json 200/500
/api/file/:id - put/delete - update/delete a file returns json 200/500

I'll post a link to the repo once I do the initial push. Repo includes a very simple index.html as demo along with ClientAPI.js.

ClientAPI.js creates a globalThis.__api with `__api.file.(get/create/update/delete)`  methods

**Comment by antelle** [comment]
@MichaelLeeHobbs next thing you need to do, is build a config on this server and provide it to KeeWeb, you can find more about configs here: https://github.com/keeweb/keeweb/wiki/Configuration#json-app-config

It should be something like:
```json
{
  "settings": {
    "webdavSaveMethod": "put"
  },
  "files": [
    { "storage": "webdav", "name": "xxx", "path": "url" }
  ]
}
```

Give it the url from one of files and see if this works. It should download / upload the file to its place.

> ClientAPI.js

This is not needed for KeeWeb, it won't load this API

**Comment by MichaelLeeHobbs** [comment]
@antelle will take a look at the info above latter.

https://github.com/MichaelLeeHobbs/keeweb-server

**Comment by zoda61** [comment]
Hello all of you!

I'm not a programmer.

I have spent weeks now to find the sleekest WebDAV server solution for shared hosted web server based on Apache. The native build in WebDAV server in Apache is not activated and implemented for web hotel users. Web hotel with cPanel support has implemented Web Disk as theirs WebDAV server. Users can define users and password etc. I do not know what WebDAV methods are supported. WebDAV methods (COPY, LOCK, MKCOL, MOVE, PROPFIND, PROPPATCH, UNLOCK). Wen hotels without cPanel like DirectAdmin panel usually not supports WebDAV server implementation out of the box. 

I could not find the complete list what WebDAV methods KeeWeb using with multiuser set up in mind. Owncloud, Nextsloud feels big packages, FileRun is less, but I could not get answer what WebDAV methods are implemented. The next step was looking for Wordpress plugins for file sharing or private cloud, but no one has WebDAV server. Wordpress is very big now and has user account manager a lot, backup functions, file shares, and so on. For me, who works even as webdesigner and always looking for possibilities to integrate useful functions, the KeeWeb could be perfect for multiuser setup integrated into Wordpress. So I agree, I'd like to use a KeeWeb including a ready to go WebDAV server with all of the functions and WebDAV methods needs for multiuser support and even one alternative KeeWeb Wordpress plugin (including the WebDAV server) integrated with Wordpress User Management for easier setup and administration. My aim is to let Wordpress Users first log into theirs Wordpress site (using for ex Ultimate Member plugin) and only after successful login (can be 2FA too) get access to theirs KeeWeb url's to connect to theirs kdbx database on the server. Any tips for a sleek WebDAV server for shared apache hosting appreciated. :-)

**Comment by KaKi87** [comment]
Hello,

I am strongly against a back-end for KeeWeb, because the ability to work without back-end is one of its strongest benefits.

Additionally, [WebDAV is easy to enable using Apache](https://git.kaki87.net/KaKi87/tutorials/src/branch/master/how_to_self_host_keeweb/README.md).

However, I would like to suggest allowing to automatically use the KDBX file's name and password as WebDAV username and password.

Thanks

**Comment by antelle** [comment]
@KaKi87 if you use Kdbx file name as WebDAV password, it becomes easier to crack this password... the password is stored somewhere on server, and vulnerabilities there (for example, old WebDAV server) may expose the KDBX file. So, not sure if it worth recommending this.

**Comment by Coises** [comment]
@KaKi87 Alas, the option to enable WebDAV in the web server is not typically available to those of us using shared hosting.

As mentioned, I have all of NextCloud installed just to get the WebDAV functionality... swatting a fly with a sledgehammer... but I could find no simple, easy-to-install-on-shared-hosting implementation of just WebDAV.

One of the reasons I thought of a custom back end, rather than just a WebDAV package, is the password issue. I'd be surprised if the password would be safe from server compromise using WebDAV (but I don't know that for certain). A custom implementation could make sure to use a form of "zero-knowledge proof" such as challenge-response using a public/private key pair where the private key, generated from the password, never leaves the client.

I agree that a _custom_ back end absolutely should not be required — but the requirement for some "back end," be it a proprietary one like Dropbox or an open one like WebDAV, is unavoidable whenever the kdbx file is stored remotely. I'm just asking for one that's easy to install on shared hosting.

If there is a way (and I don't know that there is) using something like WebAssembly to make SFTP accessible from client-side code in current browsers, that would be an excellent alternative solution, since almost every hosting service makes the hosted files available for management using SFTP.

**Comment by zoda61** [comment]
I'd like to recommend the self-hosted FileRun with own WebDAV server for shared hosting. Written in PHP.
Free license for up to 10 users. https://filerun.com/

**Comment by KaKi87** [comment]
> the password is stored somewhere on server

Apache `htpasswd` uses bcrypt.
> I'd like to recommend the self-hosted FileRun

NextCloud may be overkill in comparison to FileRun, but it's easier to install just because of being open source.
